x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/cache/CacheService.java

```diff
@@ -338,82 +351,130 @@ public void removeFromCache(final CacheKey cacheKey) {
-     * @param runnable      a runnable to execute
+     * @param shardEviction the shard eviction to process
      */
-    private void runIfShardMarkedAsEvictedInCache(ShardEviction shardEviction, Runnable runnable) {
-        try (Releasable ignored = shardsEvictionLock.acquire(shardEviction)) {
-            boolean success = false;
+    void processShardEviction(ShardEviction shardEviction) {
+        assert isPendingShardEviction(shardEviction) : "shard is not marked as evicted: " + shardEviction;
+        assert assertGenericThreadPool();
+
+        shardsEvictionsLock.readLock().lock();
+        try {
             try {
-                if (evictedShards.remove(shardEviction)) {
-                    runnable.run();
+                final List<CacheFile> cacheFilesToEvict = new ArrayList<>();
+                cache.forEach((cacheKey, cacheFile) -> {
+                    if (shardEviction.matches(cacheKey)) {
+                        cacheFilesToEvict.add(cacheFile);
+                    }
+                });
+                for (CacheFile cacheFile : cacheFilesToEvict) {
+                    try {
+                        cache.invalidate(cacheFile.getCacheKey(), cacheFile);
+                    } catch (RuntimeException e) {
+                        logger.warn(() -> new ParameterizedMessage("failed to evict cache file {}", cacheFile.getCacheKey()), e);
+                        assert false : e;
+                    }
                 }
-                success = true;
+                logger.debug("shard eviction [{}] processed with [{}] cache files invalidated", shardEviction, cacheFilesToEvict.size());
             } finally {
-                if (success == false) {
-                    final boolean added = evictedShards.add(shardEviction);
-                    assert added : shardEviction;
+                synchronized (shardsEvictionsMutex) {
+                    final Future<?> removedFuture = pendingShardsEvictions.remove(shardEviction);
+                    assert removedFuture != null;
                 }
             }
+        } finally {
+            shardsEvictionsLock.readLock().unlock();
+        }
+    }
+
+    /**
+     * Processes and waits for all pending shard evictions to complete.
+     */
+    private void processAllPendingShardsEvictions() {
+        final List<Future<?>> pendingFutures;
+        synchronized (shardsEvictionsMutex) {
+            allowShardsEvictions = false;
+            pendingFutures = new ArrayList<>(pendingShardsEvictions.values());
+        }
+        if (pendingFutures.isEmpty() == false) {
+            try {
+                if (shardsEvictionsLock.writeLock().tryLock(10L, TimeUnit.SECONDS) == false) {
+                    logger.warn("timeout when waiting for [{}] shards evictions to be processed", pendingFutures.size());
+                    shardsEvictionsLock.writeLock().lock(); // wait indefinitely
+                }
+                pendingFutures.forEach(FutureUtils::get);
```

Comment:
I think this can deadlock, since when we hold the write lock, a job that did not start yet cannot get the readlock?

But if we early terminate the submitted job, we should not need this, since the execution will do nothing?

Reply:
+1 and +1